### ISSUE TITLE
lambdify does not play well with piecewise expressions

### ISSUE THREAD

**Issue by sergiocallegari** [sympy/sympy#28585]
When you call the function obtained by lamdify with vector arguments, the expressions in *all* the branches of the original piecewise expressions get computed *for all values*, including those that should be excluded by the conditions attached to the branches. This causes exceptions or warnings.

For instance, if a branch contains the computation of a square root, then that branch may get computed even for values that cause the square root argument to be negative *even if the condition attached to the branch says that that branch expression should not be used for those values*.

I will try to build a minimal example of the problem shortly.

**Comment by jhaayushkumar** [comment]
@sergiocallegari  should i work on it.

**Comment by oscarbenjamin** [comment]
Here is an example:
```python
In [1]: p = Piecewise((sqrt(x), x >= 0), (sqrt(-x), True))

In [2]: p
Out[2]: 
⎧  √x    for x ≥ 0
⎪                 
⎨  ____           
⎪╲╱ -x   otherwise
⎩                 

In [3]: f = lambdify(x, p, 'numpy')

In [4]: import inspect

In [5]: print(inspect.getsource(f))
def _lambdifygenerated(x):
    return select([greater_equal(x, 0),True], [sqrt(x),sqrt(-x)], default=nan)


In [6]: f(2)
<lambdifygenerated-1>:2: RuntimeWarning: invalid value encountered in sqrt
  return select([greater_equal(x, 0),True], [sqrt(x),sqrt(-x)], default=nan)
Out[6]: array(1.41421356)
```
The question is what numpy code should be used here if not `np.select`.

**Comment by bjodah** [comment]
We could have `lambdify` create helper functions for each `Piecewise`, something along the lines of:
```python
>>> def _piecewise01(x):
...     x = asarray(x)
...     out = zeros_like(x)*nan
...     mask = greater_equal(x, 0)
...     sqrt(x, out=out, where=mask)
...     sqrt(-x, out=out, where=~mask)
...     return out
... 

>>> 

>>> _piecewise01([1.2, -1.2])
array([1.09544512, 1.09544512])
```

**Comment by sergiocallegari** [comment]
Wouldn't numpy's piecewise help here?

In:

```
def f(x):
    return np.piecewise(x, (x>0, x<0), (lambda x:np.sqrt(x) , lambda x:np.sqrt(-x)))
```

each lambda appears to gets called only for the appropriate input values.

**Comment by bjodah** [comment]
@sergiocallegari I think that would work. Searching for "lambdify piecewise select" in issues here turns up a bunch of relevant issues, the one with the most thorough discussion seems to be:
- https://github.com/sympy/sympy/issues/15014

It looks like `numpy.piecewise` was used successfully. So likely "all" we need is for someone to write up a PR. I do get the impression that there might be some jumping through hoops involved to retain current broadcasting behavior.

**Comment by vortex-wq** [comment]
Hi Bjorn and @sergiocallegari , I explored replacing `np.select` with `np.piecewise`, but it doesn't appear to be a drop-in replacement:

- `np.piecewise` requires mutually exclusive `boolean` masks rather than a final `True` default branch.
- It relies on callables for lazy evaluation.
- `_print_Piecewise` currently only returns a single expression string, so it can't emit helper functions or multi-statement code.

Implementing the helper-function approach suggested earlier would require changes in `lambdify` code generation rather than only in `sympy/printing/numpy.py`. What should be to approach to proceed then after reading the #15014 issue as well, I was working on extending `lambdify` but require design changes.

**Comment by tanmayBeginsJourney** [comment]
@vortex-wq are you still working on this? If not I'd like to pick it up.

What I'd try keeps the printed code a single expression, so lambdify doesn't need the multi-statement support discussed in #15014. `NumPyPrinter._print_Piecewise` prints a call to a small helper that lambdify puts in the namespace:

```python
_piecewise((x, y), [lambda x, y: x > 0, lambda x, y: True],
                   [lambda x, y: log(x)*y, lambda x, y: y])
```

The helper broadcasts the args once, then for each branch takes the elements whose first true condition is that branch and evaluates the branch only on those (`out[m] = expr(*[a[m] for a in args])`). This is @bjodah's masking idea without needing multiple statements in the generated function.

I tried it outside SymPy on @oscarbenjamin's example, `Piecewise((1/x, Ne(x, 0)), (0, True))`, a two-argument one with `x.shape == (4,)` and `y.shape == (2, 1)`, a nested Piecewise and one with no `True` branch. It gave the same values and shapes as `select`, with no warnings. Timing against the current `select` code: 20–60% slower for a scalar or an array up to 10^4 elements (7 → 11 µs for a scalar), about the same at 10^6 elements.

Before writing a PR I'd like to know what you think about:

- `pycode()` / `NumPyPrinter` output used outside lambdify would call a `_piecewise` that doesn't exist there. Should this be a printer setting that lambdify turns on, or lambdify only?
- The result dtype would depend on which branches are hit, where `select` promotes over all of them (e.g. a complex branch that's never taken gives a float result).
- It can't work under numba or `jax.jit`, so the JAX/CuPy printers would keep `select`. I'm not sure if the overhead on small inputs is acceptable, perhaps it can be made opt-in instead?
